### PR TITLE
feat(261): Adds parameter to descend into lists when scrubbing None p…

### LINKED ISSUE #261: Add descend in to list option for 'scrub_none_parameters'
##### SUMMARY
During the implementation of the `scrub_none_parameters` helper function we [talked about](https://github.com/ansible-collections/amazon.aws/pull/184) whether a "descend in to lists" option was necessary.  For the sake of getting it merged, the option was omitted.

Now I have sat down to start to try to use the helper function it has become clear that we definitely need the option and so this is why this issue is being raised.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
scrub_none_parameters helper function

##### ADDITIONAL INFORMATION
As `scrub_none_parameters` has not yet been implemented anywhere I am going to make the change to descend in to lists y default.

### PR DESCRIPTION
**Pull request by wimnat**
…arams

##### SUMMARY
Fixes #261 

Adding parameter to descend in to lists when scrubbing None parameter

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
scrub_none_parameters helper function (core)

##### ADDITIONAL INFORMATION
Defaulting to True as not yet implemented anywhere

### CODE REVIEW

File: plugins/module_utils/core.py
Comment (tremble) [inline plugins/module_utils/core.py:380]: Although I don't think boto/boto3 generally use them this won't handle lists of lists...

Not a blocker, just thinking out loud.

File: tests/unit/module_utils/core/test_scrub_none_parameters.py
Comment (tremble) [inline tests/unit/module_utils/core/test_scrub_none_parameters.py:94]: I think we'd get better coverage if this was 

```
@pytest.mark.parametrize("input_params, output_params_descend, output_params_nodescend", scrub_none_test_data) 
def test_scrub_none_parameters(input_params, output_params_descend, output_params_nodescend):
    assert scrub_none_parameters(input_params) == output_params_descend
    assert scrub_none_parameters(input_params, True) == output_params_descend
    assert scrub_none_parameters(input_params, False) == output_params_nodescend
```
a) We test the default behaviour - will cause the unit tests to flag things up if someone changed the defaults
b) We test that the descend code didn't have some weird side effect we didn't expect

**Review by tremble — CHANGES_REQUESTED** [review]

File: plugins/module_utils/core.py
Comment (felixfontein) [inline plugins/module_utils/core.py:378]:
```suggestion
            clean_parameters[k] = scrub_none_parameters(v, descend_into_lists=descend_into_lists)
```
Shouldn't it passed on recursively? In any case, the documenation should be extended, because it doesn't mention recursion *at all*.

File: changelogs/fragments/261-scrub-params-descend-into-lists.yml
Comment (felixfontein) [reply changelogs/fragments/261-scrub-params-descend-into-lists.yml:4]: Since also module_utils are part of the public API of a collection, this is a breaking change as it could have been used in a third-party collection/module/plugin. Better set the default value to `False` to avoid the hassle of having to wait for a new major release.

Comment (jillr) [reply changelogs/fragments/261-scrub-params-descend-into-lists.yml:4]: We'll have a major release in the next few months, so we could plan to change the default at that time.

**Review by jillr — COMMENTED** [review]
Hi @winmat, thanks very much for the work starting this. Are you able to continue working on this PR or would you mind if someone continued with it?  This would solve a few bugs so I'd like to be sure we can get it released asap.

File: changelogs/fragments/261-scrub-params-descend-into-lists.yml
Comment (tremble) [reply changelogs/fragments/261-scrub-params-descend-into-lists.yml]: Once we get this merged I'll prep a 'breaking' change which we can merge once the decision is made that the next release is 2.0.0

**Review by tremble — APPROVED** [review]

**Review by felixfontein — APPROVED** [review]

**Review by jillr — APPROVED** [review]

**Comment by jillr** [comment]
@tremble I'm happy for this to merge once CI is green, reran it a few times

**Comment by felixfontein** [comment]
Hmm, according to Zuul, the stable-2.9 pylint sanity test is failling:
```
2021-03-09 12:12:39.943455 | centos-8 | ERROR: plugins/modules/aws_s3.py:342:11: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.943664 | centos-8 | ERROR: plugins/modules/aws_s3.py:345:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.943868 | centos-8 | ERROR: plugins/modules/aws_s3.py:377:11: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.944099 | centos-8 | ERROR: plugins/modules/aws_s3.py:382:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.944305 | centos-8 | ERROR: plugins/modules/aws_s3.py:407:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.944508 | centos-8 | ERROR: plugins/modules/aws_s3.py:458:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.944711 | centos-8 | ERROR: plugins/modules/aws_s3.py:487:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.944912 | centos-8 | ERROR: plugins/modules/aws_s3.py:555:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.945126 | centos-8 | ERROR: plugins/modules/aws_s3.py:580:11: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.945326 | centos-8 | ERROR: plugins/modules/ec2_ami.py:581:20: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.945528 | centos-8 | ERROR: plugins/modules/ec2_ami.py:674:20: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.945730 | centos-8 | ERROR: plugins/modules/ec2_ami_info.py:260:16: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.945932 | centos-8 | ERROR: plugins/modules/ec2_eni_info.py:224:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.946150 | centos-8 | ERROR: plugins/modules/ec2_key.py:177:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.946350 | centos-8 | ERROR: plugins/modules/s3_bucket.py:269:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.946552 | centos-8 | ERROR: plugins/modules/s3_bucket.py:300:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.946753 | centos-8 | ERROR: plugins/modules/s3_bucket.py:323:12: duplicate-except: Catching previously caught exception type ClientError
2021-03-09 12:12:39.946955 | centos-8 | ERROR: plugins/modules/s3_bucket.py:358:12: duplicate-except: Catching previously caught exception type ClientError
```

**Pull request merged**
